electrumsv/network_support/esv_client_types.py

```diff
@@ -0,0 +1,159 @@
+        return {"error": {"reason": self.reason,
+                          "status_code": self.status}}
+
+    @classmethod
+    def from_websocket_dict(cls, message: Dict[str, WebsocketError]) -> 'Error':
+        reason = message["error"]["reason"]
+        status = message["error"]["status_code"]
+        return cls(reason, status)
+
+    def __str__(self) -> str:
+        return f"Error(reason={self.reason}, status={self.status})"
+
+# ----- HeaderSV types ----- #
+class HeaderResponse(TypedDict):
+    hash: str
+    version: int
+    prevBlockHash: str
+    merkleRoot: str
+    creationTimestamp: int
+    difficultyTarget: int
+    nonce: int
+    transactionCount: int
+    work: int
+
+
+class TipResponse(TypedDict):
+    header: HeaderResponse
+    state: str
+    chainWork: int
+    height: int
+
+
+# ----- Peer Channel Types ----- #
+class TokenPermissions(IntFlag):
+    NONE            = 0
+    READ_ACCESS     = 1 << 1
+    WRITE_ACCESS    = 1 << 2
+
+
+class PeerChannelToken(NamedTuple):
+    permissions: TokenPermissions
+    api_key: str
+
+
+# Todo - we may need to persist a mapping of channel_id -> PeerChannelType
+#  Otherwise we cannot know how to parse the received messages.
+#  E.g. We need to know that a MAPICallbackResponse type json structure will be received
+#  in advance to know how to parse and process it.
+class PeerChannelType(IntFlag):
+    NONE            = 0
+    MERCHANT_API    = 1 << 1
+    BACKUP          = 1 << 2
+
+
+ChannelId = str
+GenericJSON = Dict[Any, Any]
+
+
+# ViewModel refers to json response structures
+class RetentionViewModel(TypedDict):
+    min_age_days: int
+    max_age_days: int
+    auto_prune: bool
+
+
+class PeerChannelAPITokenViewModelGet(TypedDict):
+    id: int
+    token: str
+    description: str
+    can_read: bool
+    can_write: bool
+
+
+class PeerChannelViewModelGet(TypedDict):
+    id: str
+    href: str
+    public_read: bool
+    public_write: bool
+    sequenced: bool
+    locked: bool
+    head_sequence: int
+    retention: RetentionViewModel
+    access_tokens: List[PeerChannelAPITokenViewModelGet]
+
+
+class APITokenViewModelGet(TypedDict):
+    id: str
+    token: str
+    description: str
+    can_read: bool
+    can_write: bool
+
+
+class MAPICallbackResponse(TypedDict):
+    callbackPayload: str
+    apiVersion: str
+    timestamp: str
+    minerId: str
+    blockHash: str
+    blockHeight: int
+    callbackTxId: str
+    callbackReason: str
+
+
+# These are both for json but they represent an
+# underlying json vs binary payload
+class MessageViewModelGetJSON(TypedDict):
+    sequence: int
+    received: str
+    content_type: str
+    payload: MAPICallbackResponse  # Later this will be a Union of multiple message types
+
+
+class MessageViewModelGetBinary(TypedDict):
+    sequence: int
+    received: str
+    content_type: str
+    payload: str  # hex
+
+
+PeerChannelMessage = Union[MessageViewModelGetJSON, MessageViewModelGetBinary]
```

Comment:
Dump JSON and replace typed dict for namedtuple for binary.

Reply:
Do you realise that MessageViewModelGetBinary is also json but for wrapping a binary message (and therefore the payload is hex)...

If we want an octet-stream response we need to forge our own path and deviate from the Peer Channel reference implementation.

Reply:
I'm not opposed to doing our own custom binary protocol but either way mAPI callbacks are JSON so for the purposes of this PR, I don't see much to be gained and it's not exactly a huge feature to change at a later date - might take half a day or so.